exporter/opencensusexporter/opencensus.go

```diff
@@ -129,22 +180,52 @@ func OpenCensusTraceExportersFromViper(v *viper.Viper) (tps []consumer.TraceCons
 	}
 
 	tps = append(tps, oexp)
+	doneFns = append(doneFns, oce.stop)
 
 	// TODO: (@odeke-em, @songya23) implement ExportMetrics for OpenCensus.
 	// mps = append(mps, oexp)
 	return
 }
 
+func (oce *ocagentExporter) stop() error {
+	wg := &sync.WaitGroup{}
+	var errors []error
+	var errorsMu sync.Mutex
+	visitedCnt := 0
+	for currExporter := range oce.exporters {
+		wg.Add(1)
+		go func(exporter *ocagent.Exporter) {
+			defer wg.Done()
+			err := exporter.Stop()
+			if err != nil {
+				errorsMu.Lock()
+				errors = append(errors, err)
+				errorsMu.Unlock()
+			}
+		}(currExporter)
+		visitedCnt++
+		if visitedCnt == cap(oce.exporters) {
+			// Visited and started Stop on all exporters, just wait for the stop to finish.
+			break
+		}
+	}
+
+	wg.Wait()
+
+	return internal.CombineErrors(errors)
+}
+
 func (oce *ocagentExporter) PushTraceData(ctx context.Context, td data.TraceData) (int, error) {
-	// Get an exporter worker round-robin
-	exporter := oce.exporters[atomic.AddUint32(&oce.counter, 1)%uint32(len(oce.exporters))]
+	// Get first available exporter.
+	exporter := <-oce.exporters
```

Comment:
I think we have a race condition here. If `stop()` is called after this line but before line 228, the `exporter` that we have just poped from `oce.exporters` will not be stopped.

Reply:
that is not the case because `range` over channels just ends if the channel is closed but that never happens, so we leave the loop by counting the number of exporters (the capacity of the channel). Eventually this exporter is returned to the channel and the range picks it up to stop it.

Reply:
OK, makes sense, I didn't notice the counting logic.

How about the situation when stop() finishes and PushTraceData() tries to read from `oce.exporters`? It will get stuck because the channel is open but there is nothing to read from it.

Reply:
It will be stuck but this doesn't prevent shutdown of the app, but, I'm not considering the future scenario in which we may want to shutdown only the pipeline. We should define some behavior here (it seems more natural to require the workers to stop submitting data before asking for a stop of the exporter), that said, returning an error in this case is the correct approach, I will add that to the code.